File: 3434.maximum-frequency-after-subarray-operation.py
#
# @lc app=leetcode id=3434 lang=python3
#
# [3434] Maximum Frequency After Subarray Operation
#

# @lc code=start
from typing import List

class Solution:
    def maxFrequency(self, nums: List[int], k: int) -> int:
        n = len(nums)
        total_k = nums.count(k)
        max_gain = 0
        for t in range(1, 51):
            # Kadane on the fly
            if n == 0:
                continue
            score = (1 if nums[0] == t else 0) - (1 if nums[0] == k else 0)
            max_ending = score
            max_so_far = score
            for i in range(1, n):
                score = (1 if nums[i] == t else 0) - (1 if nums[i] == k else 0)
                max_ending = max(score, max_ending + score)
                max_so_far = max(max_so_far, max_ending)
            max_gain = max(max_gain, max_so_far)
        return total_k + max_gain

# @lc code=end

File: 3434.maximum-frequency-after-subarray-operation.cpp
#include <bits/stdc++.h>
using namespace std;

/*
 * @lc app=leetcode id=3434 lang=cpp
 *
 * [3434] Maximum Frequency After Subarray Operation
 */

// @lc code=start
class Solution {
public:
    int maxFrequency(vector<int>& nums, int k) {
        int totalK = 0;
        for (int x : nums) if (x == k) totalK++;

        int bestGain = 0;
        // Try converting each possible value v into k by choosing x = k - v
        for (int v = 1; v <= 50; v++) {
            if (v == k) continue;
            int cur = 0, best = 0;
            for (int a : nums) {
                int delta = 0;
                if (a == v) delta = 1;
                else if (a == k) delta = -1;
                cur = max(0, cur + delta);
                best = max(best, cur);
            }
            bestGain = max(bestGain, best);
        }
        return totalK + bestGain;
    }
};
// @lc code=end
